### DISCUSSION TITLE
Dashboard priorities

### DISCUSSION THREAD

**Discussion by bkeepers** [bensheldon/good_job#571] [Ideas]
:wave: I am interested in working on some improvements to the dashboard. Before I get too far, I thought I would ask for feedback from @bensheldon and the rest of the community.

First, here are open issues that are related to the dashboard:

* https://github.com/bensheldon/good_job/issues/438
* https://github.com/bensheldon/good_job/issues/408
* https://github.com/bensheldon/good_job/issues/543
* https://github.com/bensheldon/good_job/issues/547
* https://github.com/bensheldon/good_job/issues/540
* https://github.com/bensheldon/good_job/issues/446
* https://github.com/bensheldon/good_job/issues/424

What else would you like to see improved?

**Comment by bkeepers** [comment]
Can the "Jobs" & "Executions" tabs be unified? I find the two separate tabs to be a little awkward. Looking at the internals, I see why they're presented that way, but I wonder if it would make sense to unify them in the dashboard. Or is there value that I don't understand in being able to browse and filter them separately?

**Reply by bensheldon** [reply]
I think the "Executions" tab should go away. 

I think maybe the chart should remain (a separate charts tab?) but i think tabular view should go away. 

The one weirdness, as an fyi, is search. For example, an operator might want to search for an error message that lives on an past execution. Something to tackle later.

**Reply by bkeepers** [reply]
> I think the "Executions" tab should go away.

Ok, cool. That's in line with what I was thinking too. The only thing I like about the executions page is the "Errors" filter, but I haven't looked close enough to see if "Retried" and "Discarded" cover those.

> I think maybe the chart should remain (a separate charts tab?)…

Eventually I wonder if a landing page that is just an overview with charts, stats, etc, would make sense.

> …but i think tabular view should go away.

Can you explain what you mean by this?

**Reply by bensheldon** [reply]
>> …but i think tabular view should go away.
>
> Can you explain what you mean by this?

I mean that the "Executions" tab and the _table_ should go away. 

I think we're in agreement 🙌

**Comment by bensheldon** [comment]
Thanks for opening this discussion! I hold these ideas pretty lightly (so please probe/push-back/expand-upon) but wanted to share all of the things that are floating around in my head.

The ideal information hierarchy for Jobs as I understand it. Drilling down, top to bottom:

- Time period/window. Currently, locked to a 24-hour lookback. I think it could include both a lookback, as well as a partial look forward (e.g. what is scheduled to execute in the next 15 minutes). Ideally this would be changeable.
- State.
- Queues _or_ Job Class. It's possible that one would want to drill-down by one or the other or both. And it's possible that a single Job Class could be enqueued on multiple queues. A bit of a pickle.

The chart has always been a bit of a placeholder. It's currently the count of enqueued, split by queue name. I'm thinking that there would be two different kinds of charts:

- A drill-down chart (that occupies the current place of the chart) that simply shows the count of jobs in that state. I think it's _possible_ to customize the chart for each state, but we can defer that idea.
- A dedicated Charts page (which maybe should be the landing page for the dashboard that has a few different charts:
  - A time chart that shows enqueued, executed, and errored jobs
  - A time chart that shows {enqueued, executed, errored} broken down by {queue, job class}
  - (most excited by) a [cumulative flow chart](https://mattphilip.wordpress.com/editorialized-glossary/cumulative-flow-diagram/) that shows latency (time in queue, execution time) 
  - others?
  
The primary developer use case questions I'm hoping the Dashboard can answer:

- First-time user: Are jobs being enqueued _and executed_? Are jobs retrying as expected?
- Intermediate/Advanced: Where are there execution resource bottlenecks and how to improve queue segmentation?

**Reply by bensheldon** [reply]
In terms of priorities, I think these are the top 3:

- #446 - I should really just do this myself because I have a strong idea of how to do it.
- #547 - The bottleneck here is design. The Job permalink page is rough and I'm very open if someone wanted to improve that page.
-  #438 - As described above, I think it's less about improving the chart as is, and instead developing a separate overview page.
